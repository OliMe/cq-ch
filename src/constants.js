export const TYPE_COMMAND = 'command';

export const TYPE_QUERY = 'query';
