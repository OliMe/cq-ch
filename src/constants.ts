export const TYPE_COMMAND = 'command';

export const TYPE_QUERY = 'query';

export const IS_BROWSER = typeof window !== 'undefined';
