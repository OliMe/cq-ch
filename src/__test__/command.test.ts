import { command } from '..';
import getTransport from '../event-transport/get-transport';
import { TYPE_COMMAND } from '../constants';

jest.mock('../event-transport/get-transport', () => {
  const trigger = jest.fn();
  return jest.fn(() => ({
    trigger,
  }));
});

describe('command', () => {
  it('creates channel successfully', () => {
    expect(command(['test_type'], 'test')).toBeInstanceOf(Function);
  });
  it('creates channel unsuccessfully', () => {
    // TODO Ignore TS errors because for raw JS we need to test argument validation.
    // @ts-ignore
    const createChannelWithoutArguments = () => command();
    // @ts-ignore
    const createChannelWithFirstIncorrectArgument = () => command('not array');
    const createChannelWithFirstArgumentEmptyArray = () => command([], 'test');
    // @ts-ignore
    const createChannelWithoutSecondArgument = () => command(['test_type']);
    // @ts-ignore
    const createChannelWithSecondIncorrectArgument = () => command(['test_type'], 200);
    expect(createChannelWithoutArguments).toThrowErrorMatchingSnapshot();
    expect(createChannelWithFirstIncorrectArgument).toThrowErrorMatchingSnapshot();
    expect(createChannelWithFirstArgumentEmptyArray).toThrowErrorMatchingSnapshot();
    expect(createChannelWithoutSecondArgument).toThrowErrorMatchingSnapshot();
    expect(createChannelWithSecondIncorrectArgument).toThrowErrorMatchingSnapshot();
  });
});

describe('created by command function', () => {
  const channel = command(['first_type', 'second_type'], 'test');
  it('throws error when incorrect arguments passed', async () => {
    // TODO Ignore TS errors because for raw JS we need to test argument validation.
    // @ts-ignore
    const callChannelWithoutArguments = async () => await channel();
    // @ts-ignore
    const callChannelWithCommandWithoutTypeProperty = async () => await channel({});
    const callChannelWithCommandWithUndefinedType = async () =>
      await channel({ type: 'third_type' });
    await expect(callChannelWithoutArguments).rejects.toThrowErrorMatchingSnapshot();
    await expect(callChannelWithCommandWithoutTypeProperty).rejects.toThrowErrorMatchingSnapshot();
    await expect(callChannelWithCommandWithUndefinedType).rejects.toThrowErrorMatchingSnapshot();
  });
  it('sends command', () => {
    channel({ type: 'first_type' });
    expect(getTransport).toHaveBeenCalledWith(TYPE_COMMAND);
    const { trigger } = getTransport('command');
    expect(trigger).toHaveBeenCalledWith('first_type', {
      type: 'first_type',
      context: 'test',
      timestamp: expect.any(Number),
    });
  });
});
