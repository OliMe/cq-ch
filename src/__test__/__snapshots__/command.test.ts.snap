// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command creates channel unsuccessfully 1`] = `"Failed to execute 'command': first argument must be an Array with at least one element."`;

exports[`command creates channel unsuccessfully 2`] = `"Failed to execute 'command': first argument must be an Array with at least one element."`;

exports[`command creates channel unsuccessfully 3`] = `"Failed to execute 'command': first argument must be an Array with at least one element."`;

exports[`command creates channel unsuccessfully 4`] = `"Failed to execute 'command': second argument must be a string."`;

exports[`command creates channel unsuccessfully 5`] = `"Failed to execute 'command': second argument must be a string."`;

exports[`created by command function throws error when incorrect arguments passed 1`] = `"Failed to execute 'commandChannel': first argument must be an Object with defined property type."`;

exports[`created by command function throws error when incorrect arguments passed 2`] = `"Failed to execute 'commandChannel': first argument must be an Object with defined property type."`;

exports[`created by command function throws error when incorrect arguments passed 3`] = `"Trying to send command with type not in interface."`;
